# Exploring Used Cars from the German eBay Website

This project analyzes an ex-kaggle dataset to explore used car listings from eBay Kleinanzeigen, a classifieds section of the German eBay website. eBay is an e-commerce company which facilitates consumer-to-consumer or business-to-consumer transactions online through its website. The aim of this project is to clean the data and analyze used car listings.


## Concepts Used

* The basics of programming in Python (arithmetical operations, variables, common data types, etc.)
* Jupyter Notebook
* Working with Strings
* Object-oriented programming
* Dates and times
* List and for loops
* Conditional statements
* Dictionaries
* Functions

## Data Overview

This dataset was originally uploaded to Kaggle by user [orgeslaka](https://www.kaggle.com/orgesleka) but isn't available there anymore.It is still [availble](https://data.world/data-society/used-cars-data) at the data.world website.

[Used Car Data](https://data.world/data-society/used-cars-data): The dataset has been reduced by sampling 50,000 data points from the full dataset. It has also been dirtied to closely resemble an original scraped dataset since the version uploaded to Kaggle was cleaned already.

## Data Exploration

We will import libraries, load the dataset and look at the first couple of rows. We will also print, describe and identify columns that can help us with our anlaysis.

In [1]:
# Importing Pandas and NumPy

import pandas as pd
import numpy as np

# Autos dataset

autos = pd.read_csv('autos.csv',encoding='Latin-1')
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [2]:
# Information about our dataset

print(autos.head())
print('\n')
print(autos.info())
print('\n')
print('Number of rows:', autos.shape[0])
print('Number of columns:', autos.shape[1])

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

For the autos dataset, we have data describing 50,000 cars with 20 columns.15 columns have a string datatype and 5 columns have an integer data type. We observe that some of the columns such as `gearbox`, `model` and `notRepairedDamage` etc have missing values.  Here's a description for each of the columns from the data.world website.


* **`dateCrawled`**: When this ad was first crawled. All field-values are tkaen form this date.

* **`name`**: name of the car

* **`seller`**: Whether the seller is private or a dealer.

* **`offerType`**: The type of listing

* **`price`**: ad price to sell the car

* **`abtest`**: Whether the listing is included in an A/B test

* **`vehicleType`**: The type of the vehicle

* **`yearOfRegistration`**: Car registration year

* **`gearbox`**: Transmission type

* **`powerPS`**: The power of the car in PS

* **`model`**: Car model name

* **`kilometer`**: Kilometers driven on the car

* **`monthOfRegistration`**: The month in which the car was first registered

* **`fuelType`**: Type of fuel used by the car

* **`brand`**: Brand of the car

* **`notRepairedDamage`**: Car damage not repaired yet.

* **`dateCreated`**: eBay listing creation date

* **`nrOfPictures`**: Number of ad pictures

* **`postalCode`**: Postal code for the location of the vehicle

* **`lastSeenOnline`**: When the crawler saw this ad online

## Data Cleanup: Column Headings

We will:

* Convert the column names from [camelcase](https://en.wikipedia.org/wiki/Camel_case) to Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case) which will make them more readable and help us understand the data better
* Reword some of the column names based on column descriptions


In [3]:
# Conversion of column names

current_columns = autos.columns

print(current_columns)

update_columns = snakecase_names = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

autos.columns = update_columns

autos.head(5)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

We will continue with data exploration to identify what other data cleaning tasks need to be performed.

* We will remove all columns where all or almost all values are the same, as these don't provide any benefit to our analysis.
* We will also be converting numeric data that has been stroed as text data.

Lets start by looking at some descriptive statistics for each column.

In [4]:
# Descriptive statistics for columns

autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

In [5]:
for column in autos.columns:
    print(column," unique values:",autos[column].unique())
    print('\n')

date_crawled  unique values: ['2016-03-26 17:47:46' '2016-04-04 13:38:56' '2016-03-26 18:57:24' ...
 '2016-03-28 10:50:25' '2016-03-08 19:25:42' '2016-03-14 00:42:12']


name  unique values: ['Peugeot_807_160_NAVTECH_ON_BOARD'
 'BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik' 'Volkswagen_Golf_1.6_United'
 ... 'Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon'
 'Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+Reifen_neu_!!'
 'Fiat_500_C_1.2_Dualogic_Lounge']


seller  unique values: ['privat' 'gewerblich']


offer_type  unique values: ['Angebot' 'Gesuch']


price  unique values: ['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']


ab_test  unique values: ['control' 'test']


vehicle_type  unique values: ['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']


registration_year  unique values: [2004 1997 2009 2007 2003 2006 1995 1998 2000 2017 2010 1999 1982 1990
 2015 2014 1996 1992 2005 2002 2012 2011 2008 1985 2016 1994 1986 2001
 2018 2013 1972 1993 1988 198

Below are some comments and first impressions of each of the column.


**`date_crawled`**: Not sure if we really need the time portion.

**`name`**: Seems fine.

**`seller`**: Has only two unique values (private & gewerblich) with frequencies 4999 and 1. This column can be possibly removed as it doesn't provide much insight.

**`offer_type`**: Seems very similar to the seller column.

**`price`**: Needs to be converted into the numeric type. The top price is \$0 with frequency 1421. This is concerning, as a \$0 price for a car doesnt make sense. It may be the case that people are giving away their car for free.

**`ab_test`**: Seems to be equally distributed between the two values.

**`vehicle_type`**: Has around 44905 missing values, with 8 unique values. Can translate values into English.

**`registration_year`**: We have some registration years that are below 1900 and others as high as 9000 - this is clearly incorrrect.

**`gearbox`**: Split between the two values (manual and automatic). Has approximately 3000 missing values.Can translate values into English.

**`power_ps`** - Seems fine.

**`model`** - Some model names are numeric, not sure if that is correct. There are 2758 missing values.

**`odometer`** - The type is not numeric, so we will have to convert that. We can also change the column name to **odometer_km** and take away the km form the columns.

**`registration_month`**: Need to get rid of month 0. The type is numeric which seems correct.

**`fuel_type`**: We have 4482 missing values. Seems fine, all fuel types are in german.Can translate values into English.

**`brand`**: Seems fine.

**`unrepaired_damage`**: Has 9829 missing values. There are three unique values, **yes**, **no** and **nan** for missing vlaues.Can translate values into English.

**`ad_created`**: Not sure if we need the time format. All adds seem to be created in 2015 or 2016.

**`num_photos`**: All values are 0, it would be appropriate to get rid of this column.

**`postal_code`**: Seems fine, it's currently a numeric type, could be converted into an object type with categories. For example, we could group postal codes beginning with 7 into a category.

**`last_seen`**: Is currently an object datatype, needs to be converted into a time object plus, the years seem to be 2016 only.

## Data Cleanup: Dropping Columns

We will drop columns where all or almost all values are the same since they don't provide useful information for analysis. The two columns we will drop are `seller` and `offer_type` since they are dominated by a value which occupies a frequency of 4999 (99.9%). 

In [6]:
### Deleting columns ###

del autos['seller']
del autos['offer_type']
del autos['num_photos']

print("Remaining number of columns: ",autos.shape[1])

Remaining number of columns:  17


## Data Cleanup: Conversion Of Categorical German Words

We will be converting columns that contain German words using dictionaries. We will be using [Google Translate](https://www.google.com/search?sxsrf=ALeKk03dQU298Jxuo9BmA1efLTKwXy83dA%3A1604769644326&source=hp&ei=bNemX56GEeXK1QHRr43wBA&q=german+to+english+translation&oq=german+to+english+translation&gs_lcp=CgZwc3ktYWIQAzICCAAyAggAMgIIADICCAAyAggAMgIIADICCAAyAggAMgIIADICCAA6BAgjECc6CAgAELEDEIMBOgUIABCxAzoFCC4QsQM6CAguELEDEIMBOgIILjoHCCMQyQMQJzoLCC4QsQMQxwEQrwE6CAguEMcBEK8BOgcIABCxAxAKOgUIABDJA1C5BFi4IGDDIWgAcAB4AIABsAGIAawTkgEEMjMuNZgBAKABAaoBB2d3cy13aXo&sclient=psy-ab&ved=0ahUKEwie7N_k-PDsAhVlZTUKHdFXA04Q4dUDCAk&uact=5) to identify the English equivalent of the German words.

We will be converting the following columns: `vehicle_type`, `gear_box`, `fuel_type` and `unrepaired_damage`.

In [7]:
### Creating dictionaries ###

vehicle_type_dict = {'bus': 'bus', 'limousine': 'limousine', 'kleinwagen': 'small car', 'kombi': 'combi', 'nan' : 'nan', 'coupe':'coupe', 'suv':'suv',
'cabrio': 'convertible','andere': 'other'}
gearbox_dict = {'manuell': 'manually', 'automatik':'automatic', 'nan':'nan'}
fuel_type_dict = {'lpg':'lpg', 'benzin':'petrol', 'diesel':'diesel', 'nan':'nan', 'cng':'cng', 'hybrid':'hybrid', 'elektro':'electro','andere':'other'}
unrepaired_damage_dict = {'nein':'no', 'nan':'nan','ja':'yes'}


### Mapping new dictionaries ###

autos['vehicle_type'] = autos['vehicle_type'].map(vehicle_type_dict)
autos['gearbox'] = autos['gearbox'].map(gearbox_dict)
autos['fuel_type'] = autos['fuel_type'].map(fuel_type_dict)
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(unrepaired_damage_dict)

### printing updated columns ###

updated_cols = ['vehicle_type','gearbox','fuel_type','unrepaired_damage']

for column in updated_cols:
    print(autos[column].unique())
    print('\n')


['bus' 'limousine' 'small car' 'combi' nan 'coupe' 'suv' 'convertible'
 'other']


['manually' 'automatic' nan]


['lpg' 'petrol' 'diesel' nan 'cng' 'hybrid' 'electro' 'other']


['no' nan 'yes']




## Data Cleanup: Conversion of Text values to Numeric

For columns `price` and `odometer`, we will:

* Remove any non-numeric characters
* Convert column values to numeric
* Rename `odometer` to `odometer_km`

In [8]:
### Removing non-numeric characters ###

autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")

autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")

### Conversion to Numeric datatype ###

autos["price"] = autos["price"].astype(int)
autos["odometer"] = autos["odometer"].astype(int)

### Renaming odometer column ###

autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

In [9]:
print('Updated price column')
print('\n')
print(autos["price"].head(5))
print('\n')
print('Updated odometer_km column')
print('\n')
print(autos["odometer_km"].head(5))

Updated price column


0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64


Updated odometer_km column


0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64


Changes are implemented correctly.

Lets continue exploring these two columns, we will:

* Analyze columns for outliers
* Analyze unique values and look at the statistical summary of the columns
* Remove outliers

Lets create a function to perform the analysis outlined.

In [10]:
### Analysis Function ###

def analyze(df):
    print('\n')
    print(df.unique().shape)
    print('\n')
    print(df.describe())
    print('\n')
    print(df.value_counts().sort_index(ascending=False))

In [11]:
### Analysis of the price column ###
analyze(autos["price"])



(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             2
50            49
49             4
47             1
45             4
40             6
35             1
30             7
29             1
25             5
20             4


There are 2357 unique values.

A number of listings are below \$50, 1421 are \$0 and a handful are above \$1 million. 

Since this is ebay, we will keep listings for \$1 and above, as the bid usually starts from here and remove all listings of \$0. Listing prices steadily increase to \$350,000 amd then jump up in price astronomically.

We will count listings above \$350,000 as outliers and remove them.

In [12]:
### Removing outliers for the price column ####

autos = autos[autos["price"].between(1,350000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

Now, the min is \$1 and the max is \$350,000. The `price` seems to be mostly contained between the \$1200 - \$7500, the [Inter Quartile Range](https://en.wikipedia.org/wiki/Interquartile_range) range seems reasonable.

Lets continue with the `odometer_km` column.

In [13]:
### Analysis of the odometer_km column ###
analyze(autos["odometer_km"])



(13,)


count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
40000       815
30000       780
20000       762
10000       253
5000        836
Name: odometer_km, dtype: int64


There are 13 unique values.

The mileage seems to be rounded, which is very rare in the case of used cars, but the values seem fairly distributed. It makes sense for most cars to have more mileage since this is a bidding website, and mostcars that people put up for auction are used/have more mileage.

There seem to be no outliers.The `odometer_km` does seem to be right skewed as the [Inter Quartile Range](https://en.wikipedia.org/wiki/Interquartile_range) is between 125000 and 150000.

We will now move on to columns trhat describe the **date** in our dataset. Some dates were created by the crawler and others came form the website. We can differentiate by looking at the data dictionary.

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Lets begin by looking at the type for these columns and their formatting 

In [14]:
date_cols = ['date_crawled', 'last_seen', 'ad_created', 'registration_month', 'registration_year']

### date column's type ###   
print(autos[date_cols].dtypes)    

### date column's formatting ###
autos[date_cols][:5]

date_crawled          object
last_seen             object
ad_created            object
registration_month     int64
registration_year      int64
dtype: object


date_crawled            last_seen           ad_created  \
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00   
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00   
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00   
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00   
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00   

   registration_month  registration_year  
0                   3               2004  
1                   6               1997  
2                   7               2009  
3                   6               2007  
4                   7               2003

## Data Cleanup: Date Extraction and Analysis

We will:

* Extract the date information (first 10 characters) from `date_crawled`, `last_seen` and `ad_created`

* Convert `date_crawled`, `last_seen` and `ad_created` into numeric representation to understand them quantitatively.

* Analyze the distribution,  for all 3 variables.

In [15]:
### Extracting the first 10 charcters for each column ###

updated_col_1 = autos['date_crawled'].str.split().str[0]
updated_col_2 = autos['last_seen'].str.split().str[0]
updated_col_3 = autos['ad_created'].str.split().str[0]

### Updating columns ####

autos.pop('date_crawled')
autos.pop('last_seen')
autos.pop('ad_created')

autos = autos.assign(date_crawled = updated_col_1)
autos = autos.assign(last_seen = updated_col_2)
autos = autos.assign(ad_created = updated_col_3)


### Updated columns ###

autos[date_cols][:5]

date_crawled   last_seen  ad_created  registration_month  registration_year
0   2016-03-26  2016-04-06  2016-03-26                   3               2004
1   2016-04-04  2016-04-06  2016-04-04                   6               1997
2   2016-03-26  2016-04-06  2016-03-26                   7               2009
3   2016-03-12  2016-03-15  2016-03-12                   6               2007
4   2016-04-01  2016-04-01  2016-04-01                   7               2003

In [16]:
### Percentage counts for date_crawled ###

autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The site was crawled daily from March 5, 2016 to April 7, 2016. The listings crawled each day seem to be approximately uniform.

In [17]:
### Percentage counts for last_seen ###

autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

This shows us dates for when the crawler last saw any listings. This helps us understand if listings were removed due to cars being sold.

The last three days show a disproportionate amount of 'last seen' listings. It is very unlikely that a large number of cars were sold but more likely to do with the ending effect of the crawling period. 

In [18]:
### Percentage counts for ad_created ###

autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

There seem to be a large variety of values, 76 to be exact. Lets look at the monthly data.

In [19]:
### Percentage counts for ad_created monthly ###

autos['ad_created'].str[:7].value_counts(normalize=True, dropna=False).sort_index()

2015-06    0.000021
2015-08    0.000021
2015-09    0.000021
2015-11    0.000021
2015-12    0.000041
2016-01    0.000247
2016-02    0.001256
2016-03    0.837496
2016-04    0.160877
Name: ad_created, dtype: float64

Most values fall within the `last_seen` months (March & April). Few even date back to 9 months (June 2015).

Lets take a look at the `registration_year` column now.

In [20]:
### Analyzing registration_year ###

print(autos["registration_year"].describe())
print(autos["registration_year"].value_counts().sort_index())

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64
1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999

A car's registration year usually indicates its age. We clearly have some odd value years such as 9999 and 1000, so an option would be to remove these values.

## Data Cleanup: Removing invalid `Registration Years`

A car can't be first registered after the lisitng is seen, a registration year above 2016 is wrong and can be removed. The [article](https://www.speedyreg.co.uk/history-of-number-plates) reports that the earliest car registration was for the year 1904, so we will remove all values before that.

Before we remove rows, lets check the percentage of the data that has inaccurate values.

In [21]:
percentage = 1 - ((autos["registration_year"].between(1900,2016)).sum()/autos.shape[0])

print("Percentage of data to remove: ",percentage*100)

Percentage of data to remove:  3.8793369710697023


Given that this is less than 4% of the data, we will drop these rows.

In [22]:
### Updated autos dataframe ###

autos = autos[autos["registration_year"].between(1900,2016)]
print(autos["registration_year"].describe())

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64


The values seem correct now. It appears that most cars were registered starting in the early 1990's.

Now lets take a look at the brand column.

In [23]:
### Analyzing brand column ### 

print(autos["brand"].value_counts(normalize=True).sort_values(ascending=False))

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

German manufacturers represent the majority of the top 5 brands, 61% approximately. Since there are a lot of brands that don't have a significant percentage in listings, thus we will limit our analysis to the top ten brands.

## Data Analysis: Exploring the `Brand` Column

We will:

* Select top 10 brands we want to aggregate
* Create an empty dictionary to store our aggregate data
* Calculate the mean for each brand

In [24]:
### Extracting the top 10 brand names ###

top_brands = autos["brand"].value_counts(normalize=True).sort_values(ascending=False).head(10).index

### Creating a list of brands with averages ###

brand_average = []

for brand in top_brands:
    brand_average.append((int(autos.loc[autos['brand'] == brand,'price'].mean()),brand))
  
### Sorting and displaying brand by average price ###

sorted_brand_mean = sorted(brand_average, reverse=True)
    
for item in sorted_brand_mean:
    print(item[1], ":", item[0])

audi : 9336
mercedes_benz : 8628
bmw : 8332
volkswagen : 5402
seat : 4397
ford : 3749
peugeot : 3094
opel : 2975
fiat : 2813
renault : 2474


There seems to be three tiers of brands:
    
* Tier 1 **(Audi, Mercedez & BMX)**: These are luxury cars with the highest prices

* Tier 2 **(Volkswagen, Seat, Ford & Peugeot)**: These are mid-tier cars with the medium prices

* Tier 3 **(Opel, Fiat & Renault)***: These are lower-tier cars with the lowest  prices

This possibly explains why Volkswagen is the most popular car as it strikes a good balance between a luxury and lower-tier car. 

Now lets look at the most common models for each brand. We will:

* Filter the most common model for each of the top 10 brands.
* Calculate the average price for the most common model for each of the top 10 brands.

In [25]:
### Extracting the most common models for each of the top 10 brands ###

for brand in top_brands:
    print("Most common model for", brand)
    print(autos.loc[autos['brand'] == brand, "model"].value_counts().sort_values(ascending=False).head(1))
    print('\n')

Most common model for volkswagen
golf    3707
Name: model, dtype: int64


Most common model for bmw
3er    2615
Name: model, dtype: int64


Most common model for opel
corsa    1592
Name: model, dtype: int64


Most common model for mercedes_benz
c_klasse    1136
Name: model, dtype: int64


Most common model for audi
a4    1231
Name: model, dtype: int64


Most common model for ford
focus    762
Name: model, dtype: int64


Most common model for renault
twingo    615
Name: model, dtype: int64


Most common model for peugeot
2_reihe    600
Name: model, dtype: int64


Most common model for fiat
punto    415
Name: model, dtype: int64


Most common model for seat
ibiza    328
Name: model, dtype: int64




In [26]:
### Average price for the most common model for each of the top 10 brands ###

brand_model_combo = [('volkswagen','golf'),('bmw','3er'),('opel','corsa'),('mercedes_benz','c_klasse'),('audi','a4'),('ford','focus'),
                    ('renault','twingo'),('peugeot','2_reihe'),('fiat','punto'),('seat','ibiza')]
brand_model_average = []

### Extracting average price for brand-model ###

for combo in brand_model_combo:
    bool_price = (autos["brand"] == combo[0]) & (autos["model"] == combo[1])
    brand_model_average.append((autos.loc[bool_price,"price"].mean().round(decimals=2),(combo[0] + "-" + combo[1])))
    
sorted_brand_model_average = sorted(brand_model_average, reverse=True)

### Displaying average price for top brand-model ###
    
for item in sorted_brand_model_average:
    print(item[1], ":", item[0])
    
print('\n')
### Dispalying average price for top brands ###

for item in sorted_brand_mean:
    print(item[1], ":", item[0])

audi-a4 : 7049.14
mercedes_benz-c_klasse : 7046.85
bmw-3er : 6001.79
volkswagen-golf : 5081.58
seat-ibiza : 3782.89
ford-focus : 3509.01
peugeot-2_reihe : 3074.33
opel-corsa : 1876.51
fiat-punto : 1534.38
renault-twingo : 1441.93


audi : 9336
mercedes_benz : 8628
bmw : 8332
volkswagen : 5402
seat : 4397
ford : 3749
peugeot : 3094
opel : 2975
fiat : 2813
renault : 2474


The same trend in prices seems to be followed for top cars and top cars given most common model.

## Data Analysis: Exploring Top Cars via Mileage

We will look to understand how the top brands rank according to average mileage and if there's a link with average price.

We will:

* Calculate the average_mileage and average_price for each brand using dictionaries

* Convert both dictionaries to series objects

* Create a pandas dataframe containing `brand`, `average_price` and `average_mileage` columns.

* Print the dataframe and analyze resutls

In [27]:
### average_mileage and average_price calculation ###

average_price = {}
average_mileage = {}

for brand in top_brands:
    average_price[brand] = int(autos.loc[autos['brand'] == brand,'price'].mean())
    average_mileage[brand] = int(autos.loc[autos['brand'] == brand,'odometer_km'].mean())

### Conversion of both dictionaries to Series ###

mileage_series = pd.Series(average_mileage).sort_values(ascending=False)
prices_series = pd.Series(average_price).sort_values(ascending=False)

### Creating the DataFrame ###

brand_mean = pd.DataFrame(mileage_series,columns=['average_mileage'])
brand_mean['average_price'] = prices_series

### Displaying our DataFrame ###

print(brand_mean)

               average_mileage  average_price
bmw                     132572           8332
mercedes_benz           130788           8628
opel                    129310           2975
audi                    129157           9336
volkswagen              128707           5402
renault                 128071           2474
peugeot                 127153           3094
ford                    124266           3749
seat                    121131           4397
fiat                    117121           2813


There is very little variation when it comes to mileage for cars of all tiers compared to prices. Seems that people put up cars for listing once they've driven a certain number of kilometers.

Lets look at the average price for top brands based on `odometer_km` groups and see if average prices follow any patterns based on mileage. We will:

* Calculate the average price for each brand based on different `odometer_km` distances and add it as a column called `odoemeter_km_groups` to our dataframe.
* We will group `odometer_km` distances in 3 groups of 5000-40000, 40000-80000 and 80000-150000 as low, mid and high distances, respectively.
* Display and analyze results

In [28]:
### Extracting different odometer groups ###

# autos["odometer_km_groups"] = autos["odometer_km"]

# autos[autos["odometer_km_groups"].between(5000,40000)] = 'low'
# autos[autos["odometer_km_groups"].between(40001,80000)] = 'mid'
# autos[autos["odometer_km_groups"].between(80001,150000)] = 'high'

# odometer_group = autos["odometer_km_groups"].value_counts().sort_index()
# print(odometer_group)

